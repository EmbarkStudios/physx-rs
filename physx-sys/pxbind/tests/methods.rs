struct Functions {
    cpp: String,
    rust: String,
}

fn gen_functions(which: &str, to_emit: &'static [&str]) -> anyhow::Result<Functions> {
    let ast = pxbind::get_parsed_ast(format!("tests/data/functions/{which}"))?;

    let mut consumer = pxbind::consumer::AstConsumer::default();
    consumer.consume(&ast)?;

    let record_filter = |_rb: &pxbind::consumer::RecBinding<'_>| false;
    let enum_filter = |_eb: &pxbind::consumer::EnumBinding<'_>| false;
    let func_filter = |fb: &pxbind::consumer::FuncBinding<'_>| {
        if let Some(cname) = fb.owning_class() {
            to_emit.is_empty() || to_emit.iter().any(|te| *te == cname)
        } else {
            false
        }
    };

    let generator = pxbind::generator::Generator {
        record_filter: Box::new(record_filter),
        enum_filter: Box::new(enum_filter),
        func_filter: Box::new(func_filter),
        ..Default::default()
    };

    let cpp = {
        let mut buffer = Vec::new();
        generator.generate_cpp_functions(&consumer, &mut buffer, 0)?;
        String::from_utf8(buffer)?
    };

    let rust = {
        let mut buffer = Vec::new();
        generator.generate_rust_functions(&consumer, &mut buffer, 0)?;
        String::from_utf8(buffer)?
    };

    Ok(Functions { cpp, rust })
}

/// Verifies that reference fields are reflected in the size and layout, but aren't
/// actually publicly exposed
#[test]
fn methods_complex() {
    let mo = gen_functions("methods_complex.h", &["PxShape", "PxThreadImpl"]).unwrap();

    insta::assert_snapshot!(mo.cpp);
    insta::assert_snapshot!(mo.rust);
}

/// Verifies constructors of various kinds are properly bound, and properly
/// named to avoid clashes
#[test]
fn constructors() {
    let mo = gen_functions("constructors.h", &["PxBoxControllerDesc", "PxFilterData"]).unwrap();

    insta::assert_snapshot!(mo.cpp);
    insta::assert_snapshot!(mo.rust);
}

/// Verifies destructors are properly emitted or omitted
#[test]
fn destructor() {
    let mo = gen_functions(
        "destructors.h",
        &[
            "PxActor",
            "PxJointLimitParameters",
            "PxJointAngularLimitPair",
        ],
    )
    .unwrap();

    insta::assert_snapshot!(mo.cpp);
    insta::assert_snapshot!(mo.rust);
}
