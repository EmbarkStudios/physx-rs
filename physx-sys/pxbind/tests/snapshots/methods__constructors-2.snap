---
source: physx-sys/pxbind/tests/methods.rs
expression: mo.rust
---
extern "C" {
    pub fn PxFilterData_new(anon_param0: PxEMPTY) -> PxFilterData;

    /// Default constructor.
    pub fn PxFilterData_new_1() -> PxFilterData;

    /// Constructor to set filter data initially.
    pub fn PxFilterData_new_2(w0: u32, w1: u32, w2: u32, w3: u32) -> PxFilterData;

    /// (re)sets the structure to the default.
    pub fn PxFilterData_setToDefault_mut(self_: *mut PxFilterData);

    /// constructor sets to default.
    pub fn PxBoxControllerDesc_new_alloc() -> *mut PxBoxControllerDesc;

    pub fn PxBoxControllerDesc_delete(self_: *mut PxBoxControllerDesc);

    /// (re)sets the structure to the default.
    pub fn PxBoxControllerDesc_setToDefault_mut(self_: *mut PxBoxControllerDesc);

    /// returns true if the current settings are valid
    ///
    /// True if the descriptor is valid.
    pub fn PxBoxControllerDesc_isValid(self_: *const PxBoxControllerDesc) -> bool;

}

